Piecewise grids for spiner EOS

Yurlungur opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "fix bug in ideal gas EOS.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file.
- [ ] LANL employees: make sure tests pass both on the github CI and on the Darwin CI

If preparing for a new release, in addition please check the following:
- [ ] Update the version in cmake.
- [ ] Move the changes in the CHANGELOG.md file under a new header for the new release, and reset the categories.

Yurlungur commented:
@jhp-lanl @dholladay00 @chadmeyer this is ready for review. I think there's more one can do in this space, but IMO this is a nice start. In particular, energy and pressure near normal density are significantly better resolved with this method. 

In anticipation of enabling spiner tables to load data from another eos, I move the `Bounds` object out of `sesame2spiner` into `singularity-eos/base`, and add tests to it.

I also found and fixed the issue @pdmullen found in riot regarding extrapolation. It was a bizarre interaction between fast logs and eospac in `sesame2spiner`. Now fixed.

Yurlungur commented:
Tests triggered on re-git.

Yurlungur commented:
The core piece of this machinery lives in spiner. So the changes in singularity are just to turn that machinery on and set up reasonable grids in `sesame2spiner`.

chadmeyer reviewed (commented):
Overall, this is a good improvement.  Most of my comments are re: documentation.  

General observations:
* Is this backwards compatible with previously generated .sp5 files or not?  I assume a newly generated .sp5 file with piecewise grids will not be compatible with older versions, but I'm not sure whether that will be true the other direction because of potential metadata changes, etc.  If there is any incompatibility like that, there will need to be some kind of error telling you what happened.
* 6 regions (2T/e x 3 rho) is probably a good balance between performance and where to put resolution to cover our needs, but is it sufficiently flexible?

Yurlungur commented on doc/sphinx/src/models.rst:
```diff
@@ -1491,16 +1490,82 @@ key-value pairs. For exampe the following input deck is for air:
+decade respectively compared to the finer region.
+
+The temperature grid has two regions, a more finely spaced region at
+low temperatures and a more finely spaced region at high
```
:+1:

Yurlungur commented on doc/sphinx/src/models.rst:
```diff
@@ -1491,16 +1490,82 @@ key-value pairs. For exampe the following input deck is for air:
+
+.. code-block::
+
+  # defaults are true
```
only the booleans. Comment now written to disambiguate.

Yurlungur commented on doc/sphinx/src/models.rst:
```diff
@@ -1491,16 +1490,82 @@ key-value pairs. For exampe the following input deck is for air:
+``numrho/decade`` parameter sets the number of points per decade in
+this central refined region. The regions at lower and higher density
+have ``rhoCoarseFactorLo`` and ``rhoCoarseFactorHi`` fewer points per
+decade respectively compared to the finer region.
```
> This whole description would benefit from an image or a diagram.

That's fair. Added.

> Also, is this the best way of inputting this? Why not something like a rhominfine->rhomaxfine? I know that the same effect can be specified by adjusting quantities specified below, and maybe this is the best way for it to be internally represented, but it might be reasonable to permit a different kind of input. Just a thought there.

I agree. I chose the model I chose here because it seemed reasonable to me in the moment. Obviously `rhominfine/rhomaxfine` is a bit more flexible. I added the ability to set `rhoMinFine` and `rhoMaxFine`.

Yurlungur commented on doc/sphinx/src/models.rst:
```diff
@@ -1491,16 +1490,82 @@ key-value pairs. For exampe the following input deck is for air:
+
+.. note::
+
+  For all grid types, the only required value in an input file is the
+  matid. All other values will be inferred from the original sesame
+  database if possible and if no value in the input file is provided.
```
Table bounds and normal density are inferred. Other options have sensible-ish defaults, which are now listed.

Yurlungur commented on sesame2spiner/examples/aluminum.dat:
```diff
@@ -0,0 +1,21 @@
+#  publicly and display publicly, and to permit others to do so.
+# ======================================================================
+
+matid=3720
+name=aluminum
+rhomin = 1e-5
+Tmin = 1
```
Mainly it's here because I tested aluminum. Also I found that the minimum density/temperatures provided by the database are frequently unnecessarily low.

Yurlungur commented on sesame2spiner/generate_files.cpp:
```diff
@@ -273,34 +274,88 @@ void getMatBounds(int i, int matid, const SesameMetadata &metadata, const Params
+      params.Get("rhoFineDiameterDecades", RHO_FINE_DIAMETER_DEFAULT);
+  Real TSplitPoint = params.Get("TSplitPoint", T_SPLIT_POINT_DEFAULT);
+
+  Real rho_fine_center = rhoAnchor;
```
Done.

Yurlungur commented:
@chadmeyer I believe I have addressed all your inline comments. A few comments regarding your observations:

>     * Is this backwards compatible with previously generated .sp5 files or not?  I assume a newly generated .sp5 file with piecewise grids will not be compatible with older versions, but I'm not sure whether that will be true the other direction because of potential metadata changes, etc.  If there is any incompatibility like that, there will need to be some kind of error telling you what happened.

It's not backwards compatible, but this is a good point. I think I can make it backwards compatible, but this is a change in *spiner* not singularity-eos. I'll submit an MR there. I'd like this to go independently of that, however.

>     * 6 regions (2T/e x 3 rho) is probably a good balance between performance and where to put resolution to cover our needs, but is it sufficiently flexible?

No idea. The code infrastructure supports more regions under the hood---it's fully general---but I had to pick something to get started and adding a fully general *input* infrastructure feels like too much for a first pass.

Yurlungur commented:
> > * Is this backwards compatible with previously generated .sp5 files or not?  I assume a newly generated .sp5 file with piecewise grids will not be compatible with older versions, but I'm not sure whether that will be true the other direction because of potential metadata changes, etc.  If there is any incompatibility like that, there will need to be some kind of error telling you what happened.
> > ```
> 
> It's not backwards compatible, but this is a good point. I think I can make it backwards compatible, but this is a change in _spiner_ not singularity-eos. I'll submit an MR there. I'd like this to go independently of that, however.

@chadmeyer see this MR on spiner https://github.com/lanl/spiner/pull/92

jhp-lanl commented on doc/sphinx/src/models.rst:
```diff
@@ -1491,16 +1492,114 @@ key-value pairs. For exampe the following input deck is for air:
+
+Thus the input block for piecewise grid might look like this:
+
+.. code-block::
+
+  # Below, all right-hand-sides are set to their default values.
+  piecewiseRho = true
+  piecewiseT = true
+  piecewiseSie = true
+
+  # the fine resolution for rho.
+  numrho/decade = 350
+  # width of the fine region for rho
+  rhoFineDiameterDecades = 1.5
+  # the lower density region is 3x less refined
+  rhoCoarseFactorLo = 3
+  # the higher density region is 5x less refined
+  rhoCoarseFactorHi = 5
+
+  # the fine resolution for T
+  numT/decade = 100
+  # the point demarking the coarse and fine regions in temperature
+  TSplitPoint = 1e4
+  # it's usually wise to to not let
+  # temperature get too small in log space if you do this
+  Tmin = 1
+  # The coarser region (above the split point) is 50 percent less refined
+  TCoarseFactor = 1.5
+
+  # energy has the split point sie(rhonormal, TSplitPoint)
+  # but we may still specify the resolution
+  numSie/decade = 100
+  sieCoarseFactor = 1.5
```
So I take it that if for example we want to increase to say 10 grid regions in some dimension, we'll need to rework how this input syntax works, yes?

jhp-lanl replied:
I wrote this at the beginning of the review. The answer is yes, but I think this is a good step in the right direction.

jhp-lanl commented on doc/sphinx/src/models.rst:
```diff
@@ -1491,16 +1492,114 @@ key-value pairs. For exampe the following input deck is for air:
+.. note::
+
+  For all grid types, the only required value in an input file is the
+  matid. Table bounds and normal density will be inferred from teh
```
```suggestion
  matid. Table bounds and normal density will be inferred from the
```
typo

jhp-lanl commented on sesame2spiner/generate_files.cpp:
```diff
@@ -273,34 +274,110 @@ void getMatBounds(int i, int matid, const SesameMetadata &metadata, const Params
 
-  Real rhoAnchor = metadata.normalDensity;
-  Real TAnchor = 298.15;
+  const Real TAnchor = 298.15;
```
```suggestion
  constexpr Real TAnchor = 298.15;
```
Any reason this can't be `constexpr`? Good addition with the `const` though!

Yurlungur replied:
Nope it can be constexpr!

jhp-lanl commented on sesame2spiner/generate_files.cpp:
```diff
@@ -273,34 +274,110 @@ void getMatBounds(int i, int matid, const SesameMetadata &metadata, const Params
-  Real TAnchor = 298.15;
+  const Real TAnchor = 298.15;
+  Real rhoAnchor = params.Get("rho_fine_center", metadata.normalDensity);
+  if (std::isnan(rhoAnchor) || rhoAnchor <= 0 || rhoAnchor > 1e8) {
```
Do we encounter `NaN`s in the `normalDensity` parameter?

Yurlungur replied:
Probably not but I am concerned about what happens if it's not defined for a given material. For example, a custom material someone tries to use. That's just me being cautious.

jhp-lanl commented on sesame2spiner/generate_files.cpp:
```diff
@@ -273,34 +274,110 @@ void getMatBounds(int i, int matid, const SesameMetadata &metadata, const Params
+  }
+
+  // Piecewise grids stuff
+  bool piecewiseRho = params.Get("piecewiseRho", true);
+  bool piecewiseT = params.Get("piecewiseT", true);
+  bool piecewiseSie = params.Get("piecewiseSie", true);
```
Can these be `const`?

Yurlungur replied:
Yep :+1:

jhp-lanl commented on sesame2spiner/generate_files.cpp:
```diff
@@ -273,34 +274,110 @@ void getMatBounds(int i, int matid, const SesameMetadata &metadata, const Params
+  bool piecewiseT = params.Get("piecewiseT", true);
+  bool piecewiseSie = params.Get("piecewiseSie", true);
+
+  Real ppd_factor_rho_lo = params.Get("rhoCoarseFactorLo", COARSE_FACTOR_DEFAULT_RHO_LO);
+  Real ppd_factor_rho_hi = params.Get("rhoCoarseFactorHi", COARSE_FACTOR_DEFAULT_RHO_HI);
+  Real ppd_factor_T = params.Get("TCoarseFactor", COARSE_FACTOR_DEFAULT_T);
+  Real ppd_factor_sie = params.Get("sieCoarseFactor", COARSE_FACTOR_DEFAULT_T);
+  Real rho_fine_diameter =
+      params.Get("rhoFineDiameterDecades", RHO_FINE_DIAMETER_DEFAULT);
+  Real TSplitPoint = params.Get("TSplitPoint", T_SPLIT_POINT_DEFAULT);
+
+  Real rho_fine_center = rhoAnchor;
```
I might be missing changes below, but it looks like all of these could be `const`.

There are other reasons to advocate for lots of `const`, but I think from a readability perspective it's helpful to understand which variables may change and which are never going to change.

Yurlungur replied:
Agreed. I'll add const tags.

jhp-lanl commented on sesame2spiner/generate_files.cpp:
```diff
@@ -273,34 +274,110 @@ void getMatBounds(int i, int matid, const SesameMetadata &metadata, const Params
+  }
+  if (piecewiseSie) {
+    // compute temperature as a reasonable anchor point
+    constexpr int NT = 1;
+    constexpr EOS_INTEGER nXYPairs = 2;
+    EOS_INTEGER tableHandle[NT];
+    EOS_INTEGER tableType[NT] = {EOS_Ut_DT};
+    EOS_REAL rho[2], T[2], sie[2], dx[2], dy[2];
+    {
+      eosSafeLoad(NT, matid, tableType, tableHandle, {"EOS_Ut_DT"}, Verbosity::Quiet);
+      EOS_INTEGER eospacEofRT = tableHandle[0];
+      rho[0] = rho[1] = densityToSesame(rhoAnchor);
+      T[0] = temperatureToSesame(TAnchor);
+      T[1] = temperatureToSesame(TSplitPoint);
+      eosSafeInterpolate(&eospacEofRT, nXYPairs, rho, T, sie, dx, dy, "EofRT",
+                         Verbosity::Quiet);
+      eosSafeDestroy(NT, tableHandle, Verbosity::Quiet);
+    }
```
Does it make sense to utilize the `EOSPAC` class here directly for this lookup?

Yurlungur replied:
Hmm... I'm not sure. The EOSPAC class is a lot heavier than just what I load up here. But it would be less code. Would prefer to leave this for now if you don't mind.

There might be a rewrite down the line when `sesame2spiner` becomes a lot lighter and leverages the eospac class. So, let's wait for that.

jhp-lanl replied:
That's fine with me then. I just hate having to deal with the EOSPAC interface

jhp-lanl commented on sesame2spiner/io_eospac.hpp:
```diff
@@ -31,83 +27,16 @@
-  RegularGrid1D grid;
-  Real offset;
-};
+constexpr int NGRIDS = 3;
```
I assume it's appropriate to set this separate from the restrictions on the `spiner` end, right? I.e. if `spiner` supports 10 grids, we'd need significant code modification to support that, right?

Yurlungur replied:
Under the hood, `Spiner::DataBox` supports any number of grids, but it must be set at compiel time. `SpinerEOS` and `sesame2spiner` can have different `NGRIDS` without fear, so long as the `NGRIDS` in `SpinerEOS` is greater than or equal to the number in `sesame2spiner`. I plan to unify this more down the road.

jhp-lanl commented on singularity-eos/base/table_bounds.hpp:
```diff
@@ -0,0 +1,281 @@
+#include <ports-of-call/portability.hpp>
+#include <ports-of-call/portable_errors.hpp>
+#include <singularity-eos/base/fast-math/logs.hpp>
+#include <spiner/databox.hpp>
```
Do you actually use `DataBox` here?

Yurlungur replied:
Nope. Removed.

jhp-lanl commented on singularity-eos/base/table_bounds.hpp:
```diff
@@ -0,0 +1,281 @@
+#include <spiner/databox.hpp>
+#include <spiner/interpolation.hpp>
+
+namespace singularity {
```
Would it make sense for this to be in its own namespace? My thinking is along a few lines:
* We do this in other utility functions such as the root finder.
* `Bounds` is a generic enough name that I'm hesitant for it to live naked in the `singularity` namespace
* These utilities only apply to tabular EOS, and specifically to those tabular EOS based on `spiner`, right?

jhp-lanl replied:
You could also rename the file `spiner_table_utils` and create a `spiner_utils` namespace.

Yurlungur replied:
I am reluctant to name this `spiner_utils` as I don't think that's really what this is. I'll stick it in a `table_utils` namespace. How's that?

jhp-lanl replied:
They only apply to tabular EOS built off of spiner though, right?

Yurlungur replied:
Currently that's true... But that would put this in `SpinerEOS` as it's a util for that, not for `Spiner` as a library. 

That said, I think this is more broadly useful for, e.g., computing log bounds and things like that.

Yurlungur replied:
Also renamed file.

jhp-lanl commented on singularity-eos/base/table_bounds.hpp:
```diff
@@ -0,0 +1,281 @@
+        offset(offset), piecewise(false), linmin_(min), linmax_(max) {}
+  Bounds(OneGrid, Real min, Real max, int N, Real offset) : Bounds(min, max, N, offset) {}
+
+  Bounds(Real min, Real max, int N, bool convertToLog = false, Real shrinkRange = 0,
+         Real anchor_point = std::numeric_limits<Real>::signaling_NaN())
+      : offset(0), piecewise(true), linmin_(min), linmax_(max) {
```
It looks like several of these arguments can be `const`.

Rather than commenting on each constructor, can you make the other constructors have `const` arguments where appropriate?

Yurlungur replied:
Most of these arguments **cannot** be `const`. However, I think that is a strong reason to name the ones that are explicitly const.

jhp-lanl commented on test/test_bounds.cpp:
```diff
@@ -0,0 +1,130 @@
+SCENARIO("Bounds can compute number of points from points per decade", "[Bounds]") {
+  WHEN("We compute the number of points from points per decade") {
+    int np = Bounds::getNumPointsFromPPD(T_min, T_max, N_per_decade_fine);
+    THEN("We get the right number") { REQUIRE(np == 1800); }
```
Where does `1800` come from? Would it make sense to calculate it instead?

Yurlungur replied:
N_PER_DECADE_FINE = 200 and that T covers 9 decades. Calculated now.

jhp-lanl commented on test/test_bounds.cpp:
```diff
@@ -0,0 +1,130 @@
+    Bounds lRhoBounds(rho_min, rho_max, np, true, 0.0, rho_normal);
+    THEN("The lower and upper bounds are right") {
+      REQUIRE(std::abs(lRhoBounds.grid.min() - singularity::FastMath::log10(rho_min)) <=
+              1e-12);
```
You use `1e-12` in a lot of places for a tolerance. Two comments:
1. I would at least like this to be a `constexpr` defined parameter since it's repeated so much
2. It seems like the value might be predicated on the type of `Real`. With the idea of making things flexible, would it make sense to define
    ```c++
    constexpr REAL_TOL = std::numeric_limits<Real>::epsilon * 1e3
    ```
    instead?

I'll let this be a blanket comment for the entire file rather than pointing out every instance.

Yurlungur replied:
Done.

jhp-lanl commented on test/test_bounds.cpp:
```diff
@@ -0,0 +1,130 @@
+      REQUIRE(std::abs(bnds.grid.min() - lrmin) <= 1e-12);
+      REQUIRE(std::abs(bnds.grid.max() - lrmax) <= 1e-12);
+      REQUIRE(bnds.grid.nGrids() == 3);
+      AND_THEN(
+          "The total number of points is less than a uniform fine spacing would imply") {
```
nice check!

jhp-lanl reviewed (approved):
Approved with comments and requested changes (all pretty minor).

I think the existing tests cover this, but please ensure that the tabular EOS changes are properly covered by our existing tests.

Yurlungur commented:
Thanks, @jhp-lanl !

> I think the existing tests cover this, but please ensure that the tabular EOS changes are properly covered by our existing tests.

Yep already on that. The existing tests do cover this on re-git. So it's very important those get run before final merge.

Yurlungur commented:
@dholladay00 please review. @chadmeyer please re-review or click approve depending.

Yurlungur commented:
(tests now all passing on re-git.)

dholladay00 commented on singularity-eos/base/table_bounds.hpp:
```diff
@@ -0,0 +1,285 @@
+  // This uses real logs
+  template <typename T>
+  static int getNumPointsFromPPD(Real min, Real max, const T ppd) {
+    constexpr Real epsilon = std::numeric_limits<float>::epsilon();
```
Why specifically `float` epsilon?

Yurlungur replied:
That's a bug. Should be double.

dholladay00 commented on singularity-eos/base/table_bounds.hpp:
```diff
@@ -0,0 +1,285 @@
+
+  static void adjustForAnchor_(const Real min, Real &max, int &N,
+                               const Real anchor_point) {
+    if (min < anchor_point && anchor_point < max) {
```
What are the implications if this condition isn't met, would it make sense to error?

Yurlungur replied:
If the condition isn't met, it will just not adjust the number of points or the max. The table will still be valid, but there won't be a grid point exactly at the anchor point. I'd like `sesame2spiner` to basically always run... so I'd rather this not error out. But I'll add a warning.

dholladay00 commented on singularity-eos/eos/eos_stellar_collapse.hpp:
```diff
@@ -220,6 +220,20 @@ class StellarCollapse : public EosBase<StellarCollapse> {
+    }
+
+   private:
+    const DataBox &field_;
```
I see that this is mostly a copy/paste move, but be warned of potentially cryptic compilation errors due to const class members.

Yurlungur replied:
In this case it should be ok, and it was a copy/paste, but point taken. Will remove the `const`s.

Yurlungur replied:
This actually has to be `const` due to the `const` correctness elsewhere in the eos object.

dholladay00 commented on test/compare_to_eospac.cpp:
```diff
@@ -47,10 +47,13 @@
 #include <singularity-eos/eos/eos.hpp>
 
-using namespace Spiner;
+// using namespace Spiner;
```
remove commented out code if truly unnecessary.

Yurlungur replied:
Removed.

dholladay00 commented on singularity-eos/base/table_bounds.hpp:
```diff
@@ -0,0 +1,285 @@
+    if (convertToLog) {
+      convertBoundsToLog_(min, max, shrinkRange);
+      if (!(std::isnan(anchor_point))) {
+        anchor_point += offset;
```
is offset non-zero here?

dholladay00 commented on singularity-eos/base/table_bounds.hpp:
```diff
@@ -0,0 +1,285 @@
+  void convertBoundsToLog_(Real &min, Real &max, const Real shrinkRange = 0) {
+    // Log scales can't handle negative numbers or exactly zero. To
+    // deal with that, we offset.
+    constexpr Real epsilon = std::numeric_limits<float>::epsilon();
```
is this a bug as well @Yurlungur ?

Yurlungur replied:
Yep. Will fix.

Yurlungur commented:
Thanks @dholladay00 I think I've addressed all your comments. Can you click approve if you agree? (I will not merge until post-change tests all pass on re-git)

chadmeyer reviewed (approved):
I think it looks good now for pushing.  Good Job @Yurlungur!

Yurlungur commented:
Thanks @chadmeyer @jhp-lanl @dholladay00 for your reviews! I now plan to merge as soon as the latest commit passes tests on re-git. Speak now if you have objections. :)

jhp-lanl commented on test/test_bounds.cpp:
```diff
@@ -0,0 +1,134 @@
+#endif
+
+constexpr int NGRIDS = 3;
+using Bounds = singularity::table_utils::Bounds<NGRIDS>;
```
@Yurlungur since the `Bounds` class depends on `spiner`, can you wrap all of this in a `SINGULARITY_USE_SPINER` `#ifdef`?

I'm failing to build the tests when `SINGULARITY_USE_SPINER` is off.

I'll create an issue too.
